src/halley.jl

```diff
@@ -42,10 +42,28 @@ function SciMLBase.__solve(prob::NonlinearProblem,
                            maxiters = 1000, kwargs...)
     f = Base.Fix2(prob.f, prob.p)
     x = float(prob.u0)
-    fx = f(x)
-    # fx = float(prob.u0)
-    if !isa(fx, Number) || !isa(x, Number)
-        error("Halley currently only supports scalar-valued single-variable functions")
+    # Defining all derivative expressions in one place before the iterations
+    if isa(x, AbstractArray)
+        if alg_autodiff(alg)
+            n = length(x)
+            a_dfdx(x) = ForwardDiff.jacobian(f, x)
+            a_d2fdx(x) = ForwardDiff.jacobian(a_dfdx, x)
+            A = Array{Union{Nothing, Number}}(nothing, n, n)
+            #fx = f(x)
+        else
+            n = length(x)
+            f_dfdx(x) = FiniteDiff.finite_difference_jacobian(f, x, diff_type(alg), eltype(x))
+            f_d2fdx(x) = FiniteDiff.finite_difference_jacobian(f_dfdx, x, diff_type(alg), eltype(x))
```

Comment:
```suggestion
            f_dfdx(x) = FiniteDiff.finite_difference_jacobian(f, x, diff_type(alg), eltype(x))
            f_d2fdx(x) = reshape(FiniteDiff.finite_difference_jacobian(x->vec(f_dfdx), x, diff_type(alg), eltype(x)),n,n,n)
```

Reply:
If you `vec` the first Jacobian call, then it's x -> vector, so then the second one is just a standard Jacobian call. If you just reshape afterwards then it can be done without a memory operation into the n x n x n form. Then the allocation `A` is not required at all.